### PR TITLE
Updates to response headers for ZIP endpoints for security and functionality

### PR DESCRIPTION
**Pull request by jamalex**
### Summary

Fixes some issues we've been seeing, and proactively protects us against others.

1. Add "Content-Security-Policy" header: More aggressive protections against content being loaded from other servers (to highlight issues that might only show up offline, avoid unwanted bandwidth usage, and protect against data leakage once we add HTML5 app data access methods -- CC @kollivier).
1. Use Django decorator to block XFrameOptionsMiddleware instead of overwriting header, to avoid "Invalid 'X-Frame-Options' header encountered" JS error (which still worked, but caused unnecessary panic).
1. Add "Access-Control-Allow-Origin" header in response to OPTIONS requests: Prevent CORS errors when HTML5 apps make AJAX requests to read files from within zip.

Many (most? all?) of these changes should also be ported over [to Studio](https://github.com/learningequality/studio/blob/4ecb84457e78df35ac14c13eec7c5829c0aa0781/contentcuration/contentcuration/views/zip.py#L30) (CC @ivanistheone @jayoshih @micahscopes).

### Reviewer guidance

I made a zip file that tests all the edge cases, which I've used for doing fairly extensive cross-browser testing on BrowserStack:
[a0de2cf520524f207eed87cc7056ac6b.zip](https://github.com/learningequality/kolibri/files/2037706/a0de2cf520524f207eed87cc7056ac6b.zip)

You can overwrite one of your existing content folder zip files with that file (renaming appropriately) so as to load it in a Kolibri sandboxed iframe easily. Current filename matches the zip for ContentNode: "Channels > PhET Interactive Simulations (en) > Physics > Motion > Gravity And Orbits".

(More manual testing would also be good, on more actual content)

Screenshot of the test zip in action:

image

The main "failure" case is that in old old versions of some browsers, "Content-Security-Policy" is ignored. This just leads to it being possible to load external resources from other domains into the iframe -- doesn't compromise anything critical.

### References

Related to (but different from) https://github.com/learningequality/kolibri/issues/877 -- which is about applying Content-Security-Policy externally to the iframes, in Kolibri itself -- CC @benjaoming.

----

### Contributor Checklist

- [x] Contributor has fully tested the PR manually
- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [x] If there are any front-end changes, before/after screenshots are included
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label

### Reviewer Checklist

- [ ] Automated test coverage is satisfactory
- [ ] Reviewer has fully tested the PR manually
- [ ] PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- [ ] External dependencies files were updated (`yarn` and `pip`)
- [ ] Documentation is updated
- [ ] Link to diff of internal dependency change is included
- [ ] CHANGELOG.rst is updated for high-level changes
- [ ] Contributor is in AUTHORS.rst

### CODE REVIEW

File: kolibri/content/views.py [hunk 1]
```diff
@@ -15,6 +18,17 @@
...
 class ZipContentView(View):
 
+    @xframe_options_exempt
+    def options(self, request, *args, **kwargs):
+        """
+        Handles OPTIONS requests which may be sent as "preflight CORS" requests to check permissions.
+        """
+        response = HttpResponse()
+        response["Access-Control-Allow-Origin"] = "*"
+        response["Access-Control-Allow-Methods"] = "GET, OPTIONS"
```
Comment (ivanistheone) [inline kolibri/content/views.py]: We might also need `Access-Control-Allow-Headers`.

This is what I get for the file from earlier today, running the code from this PR:
http://35.196.143.130/learn/#/topics/c/10f41b318dc546dabe0a596aaa537a53

Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://35.196.143.130/zipcontent/213988fd537e4a608e2b86be9632d92d.zip/xml/course_config.xml. (Reason: missing token ‘x-requested-with’ in CORS header ‘Access-Control-Allow-Headers’ from CORS preflight channel).

In FF and Chrome.

**Comment by benjaoming** [comment]
Looks good, @jamalex -- you can also only enforce it with `DEBUG=True` if you want to release it with that extra caution (then of course it's only going to be useful for testing purposes, and not a security improvement).

**Comment by ivanistheone** [comment]
@jamalex Just deployed this PR to a demo server: http://35.196.143.130/ and I have good news and bad news:

The HTML app that Swapnil was working on works now: http://35.196.143.130/learn/#/topics/c/10f41b318dc546dabe0a596aaa537a53
Woo hoo... one head of the CORS dragon has been chopped off!

On the other hand, there seems to be still some issues with the GeoGebra applet:
http://35.196.143.130/learn/#/topics/c/03380b18bcac527c965c60294c675fe6 

Chrome console:
```
Uncaught DOMException: Blocked a frame with origin "null" from accessing a cross-origin frame.
    at w (http://35.196.143.130/zipcontent/3e0c56d1cf2f246e4d14af91406818c3.zip/GeoGebra/HTML5/5.0/web3d/web3d.nocache.js:4:133)
    at v (http://35.196.143.130/zipcontent/3e0c56d1cf2f246e4d14af91406818c3.zip/GeoGebra/HTML5/5.0/web3d/web3d.nocache.js:3:671)
    at m (http://35.196.143.130/zipcontent/3e0c56d1cf2f246e4d14af91406818c3.zip/GeoGebra/HTML5/5.0/web3d/web3d.nocache.js:9:7)
    at http://35.196.143.130/zipcontent/3e0c56d1cf2f246e4d14af91406818c3.zip/GeoGebra/HTML5/5.0/web3d/web3d.nocache.js:10:51
    at HTMLDocument.d (http://35.196.143.130/zipcontent/3e0c56d1cf2f246e4d14af91406818c3.zip/GeoGebra/HTML5/5.0/web3d/web3d.nocache.js:6:71)
```

Firefox console similar stack trace with message: `SecurityError: Permission denied to access property "document" on cross-origin object`


The geogebra applet is from test channel:
https://contentworkshop.learningequality.org/channels/43c08c5c7a7d5365b8a0fe5b0b7b28ad/edit/43c08c5

Download testzip using `wget 'https://contentworkshop.learningequality.org/content/storage/3/e/3e0c56d1cf2f246e4d14af91406818c3.zip' -O testzip.zip` if you want to inspect js code that's causing the issue.

**Comment by kollivier** [comment]
@ivanistheone The dependency that is causing that error shows the following URL in my Chrome: webpack:///./src/statics/turnstile-compiled.js

Is there a file like that in the package? Or is this some webpack magic that is trying to reference the file at that URL?

Also, I've noticed that web3d internally is trying to read/write cookies, which the iframe is forbidding, so that could be causing issues with loading as well.

**Comment by kollivier** [comment]
Related question, if cookies and local storage access are blocked, do HTML5 apps have any (pre-existing) options for persisting app state when used in this mode?

**Comment by rtibbles** [comment]
This error would seem to point to a clue: `sessionStorage not supportedFailed to read the 'sessionStorage' property from 'Window': The document is sandboxed and lacks the 'allow-same-origin' flag.` - would the sessionStorage be shared between the iframe and the containing frame if we allowed this, or is this just banning different scripts within the iframe affecting the iframe's session storage?

If it's the latter, then we should allow it, and this stuff should probably work.

**Comment by jamalex** [comment]
> if cookies and local storage access are blocked, do HTML5 apps have any (pre-existing) options for persisting app state when used in this mode?

Nope, don't think so.

> sessionStorage not supported

Yeah, we'll need to modify the code on-the-fly (did a bunch of this in CK-12) to remove/mock calls to things like this.

> SecurityError: Permission denied to access property "document" on cross-origin object

I faced this problem on CK-12 as well, as PLIX contain an embedded iframe to render the interactive widget thingie. There's no way to communicate directly between embedded iframes and the main html5 app as they don't have origins (would be nice if you could set "pseudo-origins" in the sandbox, but that would have other security implications).

In CK-12, I went the route of replacing it with postMessage:
https://github.com/learningequality/sushi-chef-ck12/blob/master/chef.py#L894

**Comment by jamalex** [comment]
Any changes requested in here that can be addressed by the Kolibri code itself, rather than needing HTML5 app code changes? If not, someone willing to approve?

**Comment by jamalex** [comment]
Also, in case anyone wants to do more cross-browser testing, here's an updated zip file that tests the header stuff:
[a0de2cf520524f207eed87cc7056ac6b.zip](https://github.com/learningequality/kolibri/files/2050389/a0de2cf520524f207eed87cc7056ac6b.zip)

**Comment by kollivier** [comment]
> There's no way to communicate directly between embedded iframes and the main html5 app as they don't have origins (would be nice if you could set "pseudo-origins" in the sandbox, but that would have other security implications).

What about actually giving them different origins via subdomains? e.g. instead of http://localhost:8000/zipcontent/zipid, we do http://zipid.localhost:8000/, which then gives each zip a unique origin. Then we use the same origin CORS restriction, but allow for AJAX communication with the root Kolibri domain for things like the HTML assessment API. Would this approach still allow apps to have access to content in Kolibri that they shouldn't have access to?

**Comment by ivanistheone** [comment]
@kollivier I think changing the domain would be a much more complex strategy (this PR is with fixes we hope to ship soon, i.e. with 0.10)

**Comment by ivanistheone** [comment]
@jamalex Here is a link to the demo server with both CORS HTML5 apps for testing:
http://35.196.143.130/learn/#/topics/43c08c5c7a7d5365b8a0fe5b0b7b28ad

I'm going to approve this PR because all reamining issues seem to be edge cases to be handled by cheffing.

**Review by ivanistheone — APPROVED** [review]

**Comment by indirectlylit** [comment]
thanks!

**Pull request merged**